File: ios/RNReactNativeLocationSwitch.m
#import <Foundation/Foundation.h>
#if __has_include("RCTBridgeModule.h")
#import "RCTBridgeModule.h"
#else
#import <React/RCTBridgeModule.h>
#endif


@interface RCT_EXTERN_MODULE(RNReactNativeLocationSwitch, NSObject)
RCT_EXTERN_METHOD(isLocationEnabled:
                  (RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject)
RCT_EXTERN_METHOD(changeLocationSetting)
@end

File: ios/RNReactNativeLocationSwitch.swift

//
//  NotificationPermission.swift
//  RNCheckNotificationPermission
//
//  Created by Nam Duong on 6/24/19.
//  Copyright © 2019 Facebook. All rights reserved.
//

import Foundation
import UIKit
import CoreLocation

@objc(RNReactNativeLocationSwitch)
class RNReactNativeLocationSwitch: NSObject {
    
    @objc
    func isLocationEnabled(_ resolve: @escaping RCTPromiseResolveBlock,
                                     rejecter reject: @escaping RCTPromiseRejectBlock) {
        if CLLocationManager.locationServicesEnabled() {
            switch CLLocationManager.authorizationStatus() {
            case .notDetermined, .restricted, .denied:
               resolve(false)
            case .authorizedAlways, .authorizedWhenInUse:
                resolve(true)
            }
        } else {
           resolve(false)
        }
    }
    
    @objc
    func changeLocationSetting() {
        if #available(iOS 10.0, *) {
            if !CLLocationManager.locationServicesEnabled() {
                if let url = URL(string: "App-Prefs:root=Privacy&path=LOCATION") {
                    // If general location settings are disabled then open general location settings
                    UIApplication.shared.openURL(url)
                }
            } else {
                if let url = URL(string: UIApplication.openSettingsURLString) {
                    // If general location settings are enabled then open location settings for the app
                    UIApplication.shared.openURL(url)
                }
            }
        } else {
            // Fallback on earlier versions
        }
    }
    
    @objc
    static func requiresMainQueueSetup() -> Bool {
        return true
    }
}
